### ISSUE TITLE
cannot infer type for `__E`

### ISSUE THREAD

**Issue by MarkSwanson** [jpernst/rental#10]
I seem to be stuck on what I think is a TryNewResult issue:

(type annotations needed)
if let Ok(rent_ctx) = rent_point::RentContext::try_new(
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `__E`

Are these the docs for try_new() ?
  - https://docs.rs/rental/0.4.7/rental/examples/struct.RentRef.html

If the problem/solution isn't obvious I'm happy to create and provide a small sample program that shows how to trigger this.

Any suggestions on how to resolve this error message would be greatly appreciated.

Thanks!

**Comment by jpernst** [comment]
`try_new` supports returning any custom error type from your closures, but they must all return the SAME error type. This error just means it doesn't have enough information to determine what error type to expect. How this type is determined entirely depends on your use case, so I can't offer any more specific suggestions without a code sample.

**Comment by MarkSwanson** [comment]
Ok, I'm stuck :-)
I can't find out where to put the annotations.
Also, I'm not sure where else I could/should specify the return Error?


```
    match rent_point::RentContext::try_new(
        Box::new(v),
        |v| {
            Ok(Box::new(test1(&mut v)))
        },
        |a, _v| {
            match a.next() {
                Some(s) => Ok(Box::new(s)),
                None => return Err(ErrorKind::RentContextFailed.into())
            }
        },
        |a, _a, _v| {
            match s.get_root::<point::Reader>() {
                Ok(point) => Ok(point),
                Err(e) => return Err(ErrorKind::RentContextFailed.into())
            }
        },
    ) {
        Ok(rent_ctx) => {
            rent_ctx.rent(|point: &point::Reader| {
                assert!(point.get_z() == 99, "z != 99");

            });
        },
        Err(e) => panic!("todo: {:?}", e.0)
    }
```

**Comment by jpernst** [comment]
Try removing the `.into()` calls on the errors. Those are asking it the convert the error into some other type determined by inference, so it's getting confused again. Without them, it will know exactly what error type you're using.

**Comment by MarkSwanson** [comment]
Thanks. That seems to work fine.
I'm familiar with Lisp/Clojure macros - but as soon as you wrap things in a Rust macro I'm in unfamiliar territory (for the moment).

The next step was to realize that a.next() requires mutability (I also needed v.as_mut_ptr()) so I had to switch to #[rental_mut].
Doing that changed the closures to only accept one argument. Fixing that last thing == success!

However, when I added a lifetime to the suffix I became completely stumped again.
```
        #[rental_mut]
        pub struct RentContextWrite2 {
            v: Box<Vec<u8>>,
            s: Box<S>,
            mb: Box<m::B<S>>,
            point_builder: point::Builder<'mb>,
        }
```
Using the following code:
```
rent_context_write2.ref_rent_mut(|point_builder| &mut point_builder);
```
I get tons of rustc lifetime complaints like:

the type `&'point_builder mut point::Builder<'mb>` does not fulfill the required lifetime
note: type must outlive the static lifetime

I'm confused by this. I don't know where the static lifetime requirement is coming from.

**Comment by jpernst** [comment]
`ref_rent/ref_rent_mut` can only return references that do not contain a special rental lifetime in their type signature. By adding `'mb` to the suffix, it now contains a rental lifetime, so `ref_rent_mut` will balk at it. In these cases, you'll just need to use the regular `rent_mut` method and do the work inside the closure instead of returning the reference to the outside.

**Comment by MarkSwanson** [comment]
Yeah, rent_mut() worked perfectly. Thanks!

Fwiw I looked for rent_mut() but couldn't find it in the docs. I found ref_rent_mut() in the nice examples in the git repo test/ dir.
I feel like I'm missing some docs - where are rent_mut(), ref_rent_mut() etc. documented?

Fwiw I think I now see it: you have nice examples here: rental::examples::RentMut

I didn't see a connection between the RentMut methods like rent_mut().
Now that I see it, it seems obvious - all of your rental::examples structs have the exact same Methods - and these methods are available for all rental structs.
Heh, that was kindof a big piece I was missing :-)

I'm very glad to have this working.
Thanks for helping me through it.

/me jumps for joy!
